### PR TITLE
Access to meta data

### PR DESCRIPTION
**Pull request by DonPasquale**
The FMS data API supports access to some meta data, see: 

= > https://[your_fms]/fmi/data/apidoc/#tag/metadata .

This is a proposal to make this information accessible.

The following works:
- Product info
- Layout Names
- Script Names
- Layout Metadata

Unfortunately, retrieving the names of the databases (Database Names) throws an error due to an invalid HTTP Authorization header:

>>> FileMakerError: FileMaker Server returned error 960, HTTP Authorization header invalid 

I wonder if this is caused by the fact that 'Authorization' is optional?

= > https://[your_fms]/fmi/data/apidoc/#operation/dbNames

Many thanks!

### CODE REVIEW

**Comment by DonPasquale** [comment]
One more thought: 
Currently, Database Names, Layout Names, and Script Names are returned as the complete "fms_response" from _call_filemaker .
May just a list or tuple of the respective names would be sufficient in these cases?

**Comment by DonPasquale** [comment]
I also double checked that I can access the dabase names directly via https://[your_fms]/fmi/data/v1/databases . This works just fine. Many thanks.

**Comment by DonPasquale** [comment]
With these last smaller changes it is also possible to view a list of all database names, which are accessible via the Data API. 

Unfortunately this only works if _not yet_ logged in.

If already logged in, we see an Error as already mentioned above:

>>> FileMakerError: FileMaker Server returned error 960, HTTP Authorization header invalid

In server.py, in function 'get_databasenames', if FMS is _not_ filtering databases by name, use:

>>>         response = self._call_filemaker('GET', path)

if instead FMS is filtering by name, use:

>>>         response = self._call_filemaker('GET', path, auth=(self.user, self.password))


So, I am apparently facing two issues here:

- how to "detect", if FMS is actually set to filtering databases by name or not; and  
- I am unfortunately not understanding the inner workings of requests once we are logged in to FMS already.

I would very much appreciate if you could have a look at this!

Thank you!

Best regards.

**Comment by davidhamann** [comment]
Thanks for the PR!

1.) Currently, the Authorization header is always added if `self._token` is set. The `self._headers.pop()` doesn't really do anything as the `_call_filemaker` takes care of adding the header if there's a token (by calling `self._update_token_header` itself).

I looked into the `/fmi/data/v1/databases/` endpoint. FMS doesn't expect a Bearer access token here (in case DB filtering is on), but rather a Basic auth just like when doing a login. Once the Basic auth header is set, the API always gives a good result, so I guess we could just always send it when fetching the db info.

In your current `get_databasenames` this could be fixed by removing the code to pop the header and calling `_update_token_header`, and just doing `response = self._call_filemaker('GET', path, auth=(self.user, self.password))`. This would eventually overwrite the Authorization header with the data provided in the `auth` argument (a comment there might be good as it could lead to confusion since we're adding an Authorization header, but requests adds one as well).

2.) The meta endpoint result data always seems to nested in the response object. I think it would make sense to extract that key to make it easier to access (e.g. return the contents of the `productInfo` key directly vs. returning {'productInfo': { ...}}). The same for the databases, scripts, layouts, etc.

Just returning a tuple of database names (as you mentioned above) would certainly be an option. I could imagine that this endpoint might return more data in the future, though, (maybe extended privs, open state, etc., just like the other meta endpoints) so having it as a dict might as well be fine (and more consistent when looking at the other meta endpoint results).

3.) May I suggest that we name the meta endpoints in a slightly more general way as they sometimes represent more than just names?

Maybe something like this?

```
'meta': {
    'product':      '/fmi/data/v1/productinfo',
    'databases':    '/fmi/data/v1/databases',
    'layouts':      '/fmi/data/v1/databases/{database}/layouts',
    'scripts':      '/fmi/data/v1/databases/{database}/scripts',
}
```

The detailed layout data could then be accessed via `API_PATH['meta']['layouts'] + f'/{layout_name}'`.

The functions could then be:

* `get_product_info()`
* `get_databases()`
* `get_layouts()`
* `get_scripts()`
* `get_layout(layout_name)`

Thanks again and let me know what you think.

**Comment by DonPasquale** [comment]
Thank you for explaining how auth works here. Unfortunately, get_databases still only works for me if *not* logged in and FMS database filtering is *not* active.

I also included `return response.get('key', None)`, but I still get the nested dict. Probably another beginenrs mistake...

Also, I adjusted the PR according to your suggestions. Again, unfortunately, I was not able to make `get_layout(layout_name)` to work...

Many thanks for your time!

File: fmrest/server.py [hunk 1]
```diff
@@ -627,6 +627,115 @@ def last_script_result(self) -> Dict:
...
+
+    @_with_auto_relogin
+    def get_layout(self) -> Dict:
+        """Fetches layout metadata and returns Dict instance
+
+        Parameters
+        -----------
+        none
+        """
+        path = API_PATH['meta']['layouts'] + f'/{layout}'.format(
```
Comment (davidhamann) [inline fmrest/server.py:730]: Here, the formatting must be applied to the first part:

```
        path = API_PATH['meta']['layouts'].format(
            database=self.database,
            layout=self.layout
        ) + f'/{self.layout}'
```

File: fmrest/server.py [hunk 2]
```diff
@@ -627,6 +627,115 @@ def last_script_result(self) -> Dict:
...
+
+        Parameters
+        -----------
+        none
+        """
+        path = API_PATH['meta']['product']
+
+        # There is never any auth for retrieving the producInfo, 
+        # so let's keep the Auth header 'clean'...
+        #self._headers.pop('Authorization', None)
```
Comment (davidhamann) [inline fmrest/server.py:641]: We could remove the old comments and commented code here as we won't pop the Auth header anymore.

File: fmrest/server.py [hunk 3]
```diff
@@ -627,6 +627,115 @@ def last_script_result(self) -> Dict:
...
+
+        # _call_filemaker() will reinstate the Auth header 
+        # as long as _token is set
+        response = self._call_filemaker('GET', path)
+
+        # Reset the Auth header to its previous state
+        # but _call_filemaker does this anyway...
+        #self._update_token_header()
```
Comment (davidhamann) [inline fmrest/server.py:649]: We could remove the old comments and commented code here as we won't pop the Auth header anymore.

File: fmrest/server.py [hunk 4]
```diff
@@ -627,6 +627,115 @@ def last_script_result(self) -> Dict:
...
+        Parameters
+        -----------
+        none
+        """
+        path = API_PATH['meta']['databases']
+
+        # Looking at https://[your_fms]]/fmi/data/apidoc/#operation/dbNames:
+        # We must NOT send the _token_, irrespective of the login state, 
+        # so let's keep the Auth header 'clean'...
+        #self._headers.pop('Authorization', None)
```
Comment (davidhamann) [inline fmrest/server.py:665]: We could remove the old comments and commented code here as we won't pop the Auth header anymore (also see below the old `_call_filemaker` calls).

**Review by davidhamann — CHANGES_REQUESTED** [review]
Thanks for the changes. I left a few comments in the code (mostly removing old comments and fixing the `get_layout` issue.

Regarding your remarks:

1.) Do you still get the same error? For me the current version works for both filtered and unfiltered (tested with FMS 19.4.1.41). Are you getting the same error as before? Can you post it / a trace?

2.) For me this also works fine in my demo environment. Are you sure you tested with your latest commits? If so, can you post the error / trace?

3.) I posted the fix in the in-code comments.

**Review by DonPasquale — COMMENTED** [review]
Thank you very much!

I am of course fine with all requested changes :-)

I will have to retest and see what went wrong in my testing environement. I will get back to you with any insights

**Comment by davidhamann** [comment]
Thanks for your work! In case you are able to reproduce the error with `get_databases` and/or `get_layout` again, feel free to open another issue.

**Comment by DonPasquale** [comment]
This is just to confirm that it is working now, following a:

pip install -U https://github.com/davidhamann/python-fmrest/archive/master.zip

**Pull request merged**